### PR TITLE
oem-ibm: Capacity on demand licensing support

### PR DESCRIPTION
**Pull request by jaypadath**
This commit supports:

1. Creation or update of new license interfaces and properties supplied by host
2. Clearing of license status to Unknown, during host power off transition
3. Restoration of license interfaces during reset reload and setting the license status to Unknown

Tested By:

1. Verified that dbus hosting is done by pldm
2. CoD(Capacity on demand) values are coming up in the BMC
3. Verified that reboot of BMC does not clear values

Change-Id: I9377e66753c2df7f8e1ac8b6a4ed285db69ad090

### CODE REVIEW

File: oem/ibm/libpldmresponder/file_io_type_lic.cpp [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+    std::ofstream licJsonFile(newLicJsonFilePath,
+                              std::ios::out | std::ios::binary | std::ios::app);
+    if (!licJsonFile)
+    {
+        error("License json file '{NEW_LIC_JSON}' create failed in write",
+              "NEW_LIC_JSON", newLicJsonFilePath.c_str());
+        return -1;
+    }
+
+    if (buffer != nullptr)
```
Comment (Pavithrab7) [inline oem/ibm/libpldmresponder/file_io_type_lic.cpp:121]: if(!buffer)

**Review by Pavithrab7 — APPROVED** [review]
@jaypadath Changes looks good. But let us get the enum values `pldm_license_install_status` from PDI in a seperate PR.

**Comment by jaypadath** [comment]
> @jaypadath Changes looks good. But let us get the enum values `pldm_license_install_status` from PDI in a seperate PR.

Okay will take care of that separately

File: oem/ibm/libpldmresponder/file_io.hpp [hunk 2]
```diff
@@ -488,6 +524,9 @@ class Handler : public CmdHandler
     std::unique_ptr<sdbusplus::bus::match_t>
         vmiCertMatcher;    //!< Pointer to capture the interface added signal
                            //!< for new csr string
+    std::unique_ptr<sdbusplus::bus::match::match>
```
Comment (Pavithrab7) [inline oem/ibm/libpldmresponder/file_io.hpp:527]: match_t

**Review by Pavithrab7 — CHANGES_REQUESTED** [review]

File: host-bmc/custom_dbus.cpp [hunk 3]
```diff
@@ -27,5 +27,69 @@ std::optional<std::string>
     return std::nullopt;
 }
 
+void CustomDBus::setOperationalStatus(const std::string& path, uint8_t status)
+{
+    if (operationalStatus.find(path) == operationalStatus.end())
```
[the remark this answers is not in the document]
Comment (jaypadath) [reply host-bmc/custom_dbus.cpp:32]: Done

File: host-bmc/custom_dbus.cpp [hunk 4]
```diff
@@ -27,5 +27,69 @@ std::optional<std::string>
...
+    }
+    else
+    {
+        operationalStatus.at(path)->functional(false);
+    }
+}
+
+bool CustomDBus::getOperationalStatus(const std::string& path) const
+{
+    if (operationalStatus.find(path) != operationalStatus.end())
```
[the remark this answers is not in the document]
Comment (jaypadath) [reply host-bmc/custom_dbus.cpp:51]: Done

File: host-bmc/custom_dbus.cpp [hunk 5]
```diff
@@ -27,5 +27,69 @@ std::optional<std::string>
...
+}
+
+void CustomDBus::implementLicInterfaces(
+    const std::string& path, const uint32_t& authdevno, const std::string& name,
+    const std::string& serialno, const uint64_t& exptime,
+    const sdbusplus::com::ibm::License::Entry::server::LicenseEntry::Type& type,
+    const sdbusplus::com::ibm::License::Entry::server::LicenseEntry::
+        AuthorizationType& authtype)
+{
+    if (codLic.find(path) == codLic.end())
```
[the remark this answers is not in the document]
Comment (jaypadath) [reply host-bmc/custom_dbus.cpp:66]: Done

File: host-bmc/custom_dbus.cpp [hunk 6]
```diff
@@ -27,5 +27,69 @@ std::optional<std::string>
...
+    codLic.at(path)->serialNumber(serialno);
+    codLic.at(path)->expirationTime(exptime);
+    codLic.at(path)->type(type);
+    codLic.at(path)->authorizationType(authtype);
+}
+
+void CustomDBus::setAvailabilityState(const std::string& path,
+                                      const bool& state)
+{
+    if (availabilityState.find(path) == availabilityState.end())
```
[the remark this answers is not in the document]
Comment (jaypadath) [reply host-bmc/custom_dbus.cpp:84]: Done

File: host-bmc/custom_dbus.hpp [hunk 7]
```diff
@@ -19,6 +24,13 @@ using ObjectPath = std::string;
 using LocationIntf =
     sdbusplus::server::object_t<sdbusplus::xyz::openbmc_project::Inventory::
                                     Decorator::server::LocationCode>;
+using LicIntf = sdbusplus::server::object::object<
```
[the remark this answers is not in the document]
Comment (jaypadath) [reply host-bmc/custom_dbus.hpp:27]: Done

File: host-bmc/custom_dbus.hpp [hunk 8]
```diff
@@ -19,6 +24,13 @@ using ObjectPath = std::string;
...
+    sdbusplus::com::ibm::License::Entry::server::LicenseEntry>;
+using AvailabilityIntf = sdbusplus::server::object::object<
```
[the remark this answers is not in the document]
Comment (jaypadath) [reply host-bmc/custom_dbus.hpp:29]: Done

File: host-bmc/custom_dbus.hpp [hunk 9]
```diff
@@ -19,6 +24,13 @@ using ObjectPath = std::string;
...
+    sdbusplus::xyz::openbmc_project::State::Decorator::server::Availability>;
+using OperationalStatusIntf =
+    sdbusplus::server::object::object<sdbusplus::xyz::openbmc_project::State::
```
[the remark this answers is not in the document]
Comment (jaypadath) [reply host-bmc/custom_dbus.hpp:32]: Done

File: oem/ibm/libpldmresponder/file_io.hpp [hunk 10]
```diff
@@ -354,6 +357,39 @@ class Handler : public CmdHandler
                 }
             }
         });
+        codLicenseSubs = std::make_unique<sdbusplus::bus::match::match>(
+            pldm::utils::DBusHandler::getBus(),
+            sdbusplus::bus::match::rules::propertiesChanged(codLicObjPath,
+                                                            codLicInterface),
+            [this, hostSockFd, hostEid, instanceIdDb,
+             handler](sdbusplus::message::message& msg) {
```
[the remark this answers is not in the document]
Comment (jaypadath) [reply oem/ibm/libpldmresponder/file_io.hpp:365]: Done

File: oem/ibm/libpldmresponder/file_io.hpp (hunk 2 above)
Comment (jaypadath) [reply oem/ibm/libpldmresponder/file_io.hpp:527]: Done

**Review by Pavithrab7 — APPROVED** [review]

**Comment by rfrandse** [comment]
jenkins run tests please

**Comment by jaypadath** [comment]
jenkins run tests please

File: common/utils.hpp [hunk 11]
```diff
@@ -47,35 +48,6 @@ using inventoryManager =
 constexpr auto dbusProperties = "org.freedesktop.DBus.Properties";
 constexpr auto mapperService = ObjectMapper::default_service;
 constexpr auto inventoryPath = "/xyz/openbmc_project/inventory";
-/** @struct CustomFD
```
Comment (Patel-Kamalkumar) [inline common/utils.hpp:50]: this change is part of AIO commit, why it's showing here?

File: oem/ibm/libpldmresponder/file_io.cpp [hunk 12]
```diff
@@ -45,27 +45,30 @@ struct AspeedXdmaOp
                        //!< operation, true means a transfer from BMC to host.
 };
 
-constexpr auto xdmaDev = "/dev/aspeed-xdma";
-
-int DMA::transferHostDataToSocket(int fd, uint32_t length, uint64_t address)
```
Comment (Patel-Kamalkumar) [inline oem/ibm/libpldmresponder/file_io.cpp:50]: this change is part of AIO commit, why it's showing here?

File: host-bmc/custom_dbus.cpp [hunk 13]
```diff
@@ -27,5 +27,69 @@ std::optional<std::string>
...
+    const std::string& serialno, const uint64_t& exptime,
+    const sdbusplus::com::ibm::License::Entry::server::LicenseEntry::Type& type,
+    const sdbusplus::com::ibm::License::Entry::server::LicenseEntry::
+        AuthorizationType& authtype)
+{
+    if (!codLic.contains(path))
+    {
+        codLic.emplace(
+            path, std::make_unique<LicIntf>(pldm::utils::DBusHandler::getBus(),
+                                            path.c_str()));
```
Comment (Patel-Kamalkumar) [inline host-bmc/custom_dbus.cpp:70]: c_str() not required here

File: host-bmc/custom_dbus.cpp [hunk 14]
```diff
@@ -27,5 +27,69 @@ std::optional<std::string>
...
+}
+
+void CustomDBus::setAvailabilityState(const std::string& path,
+                                      const bool& state)
+{
+    if (!availabilityState.contains(path))
+    {
+        availabilityState.emplace(
+            path, std::make_unique<AvailabilityIntf>(
+                      pldm::utils::DBusHandler::getBus(), path.c_str()));
```
Comment (Patel-Kamalkumar) [inline host-bmc/custom_dbus.cpp:88]: c_str() not required here

**Pull request closed without merging**